events/options.go

```diff
@@ -0,0 +1,129 @@
+package events
+
+import "time"
+
+// PublishOptions contains all the options which can be provided when publishing an event
+type PublishOptions struct {
+	// Metadata contains any keys which can be used to query the data, for example a customer id
+	Metadata map[string]string
+	// Payload contains any additonal data which is relevent to the event but does not need to be
+	// indexed such as structured data
+	Payload interface{}
+	// Timestamp to set for the event, if the timestamp is a zero value, the current time will be used
+	Timestamp time.Time
+}
+
+// PublishOption sets attributes on PublishOptions
+type PublishOption func(o *PublishOptions)
+
+// WithMetadata sets the Metadata field on PublishOptions
+func WithMetadata(md map[string]string) PublishOption {
+	return func(o *PublishOptions) {
+		o.Metadata = md
+	}
+}
+
+// WithPayload sets the payload field on PublishOptions
+func WithPayload(p interface{}) PublishOption {
+	return func(o *PublishOptions) {
+		o.Payload = p
+	}
+}
+
+// WithTimestamp sets the timestamp field on PublishOptions
+func WithTimestamp(t time.Time) PublishOption {
+	return func(o *PublishOptions) {
+		o.Timestamp = t
+	}
+}
+
+// SubscribeOptions contains all the options which can be provided when subscribing to a topic
+type SubscribeOptions struct {
+	// Prefix when true will return any event which has the prefix of the topic, not just exact matches
+	Prefix bool
+	// Queue is the name of the subscribers queue, if two subscribers have the same queue the message
+	// should only be published to one of them
+	Queue string
```

Comment:
Minor : I wonder if queue is the correct terminology. It probably is, but Kafka talks about consumer groups for instance. Still think it's fine but flagging in case anyone can think of something better

Reply:
I'm happy with queue, seems like most widely used term?

Reply:
Yea queue makes sense, "I subscribe to a queue".

Reply:
"topic" works for logs, streams _and_ queues